### ISSUE TITLE
sf::Text incorrect Outline

### ISSUE THREAD

**Issue by achpile** [SFML/SFML#1313]
I just found that outline calculation is incorrect for sf::Text. Screenshots and example code are in forum thread here: https://en.sfml-dev.org/forums/index.php?topic=22690

If outlineThickness is positive - outline is drawn like it is positive and negative at the same time.
If it's negative - result is completely incorrect.

**Comment by rbrich** [comment]
Hi, this looks to be a documented behaviour in SFML:

`Text::setColor`
> Setting the fill color to a transparent color with an outline
> will cause the outline to be displayed in the fill area of the text.

`Text::setOutlineThickness`
> Be aware that using a negative value for the outline
> thickness will cause distorted rendering.

Looking at the FreeType API, sf:Text uses `FT_Stroker_Set` to set the radius (=thickness), and then  `FT_Glyph_Stroke` to draw the outline. This draws both inner and outer parts of the outline ("borders"). Both using the same radius. Negative value of radius seems to be just incorrect (not documented at all).

To draw only the outer (or inner) border, there is another function: `FT_Glyph_StrokeBorder`. This would have to be used in SFML to support alpha-blended text like in your example.

**Comment by Foaly** [comment]
Hi! 
Eventhough this is documented behaviour I think there should be some changes.

I don't see any reason why the outline should draw within the bounds of a character. I mean its called outline afterall. So its probably a good idea to switch to `FT_Glyph_StrokeBorder` like @rbrich suggested. 

Also in my opinion a negative outlineThickness doesn't make much sense. Personally I would clamp the `outlineThickness` to 0.f and document that. I know some devs here don't like things like that. But I think there should be at least an assertion saying that negative values will lead to distorted rendering.
Another option would be to draw `FT_Glyph_StrokeBorder` inner border for negative thicknesses.
[Docs](https://www.freetype.org/freetype2/docs/reference/ft2-glyph_stroker.html#FT_Glyph_StrokeBorder)

**Comment by MarioLiebisch** [comment]
Why not leave the option to whoever is using it? If the outline is positive, use `FT_Glyph_StrokeBorder`, if it's negative, use the current behavior (with an absolute value)?

**Comment by Foaly** [comment]
Actually thinking about it again I wouldn't clamp the values to positive, but instead fill negative values with some actual functionality.
When we change the behaviour to only draw an outline along the outside for positive values, we can basically pick what we want do for negative values. I think there are two options: 
Either offer the old functionality for compatibility reasons, so it is still possible to acchive the same rendering as now. This style has very little logical connection to a negative thickness value.
The other option is to draw an outline along the inside of the character. I think this would add some new functionality and makes sematically the most sense.

**Comment by MarioLiebisch** [comment]
I'd certainly go with the inside approach for negative values. Makes the most sense. Don't think we'd need any legacy workaround or deprecation here, since the old approach was basically bugged anyway.

**Comment by Foaly** [comment]
I did a quick test and unforunatelly the results are a little underwhelming. Here is what the different modes look like rendered using Arial. 
<img width="790" alt="stoke_comp" src="https://user-images.githubusercontent.com/1014762/35229896-dd4f3692-ff94-11e7-9ae5-4c1b367b2393.png">
I've tried with a few different other fonts and the results all look similar. To me this looks like a bug in FreeType since especially the first one (showing only the outside outline), most definitly does not only show the outline. Same for inside = true.
FreeType is still actively developed, so I would like to write them and show them the picture. But first we should update to the most recent version 2.9 (released on 8.1.2018). I did read through the change logs (we currently use 2.5.5 from 30.12.2014), but couldn't find anything regarding this. Nevertheless its good practice. @eXpl0it3r Do you think that that's possible?

**Comment by MarioLiebisch** [comment]
How about updating your local copy and see whether something changed?

**Comment by Foaly** [comment]
Yes I can do that. Its just that I don't know what configuration FreeType has been set up for SFML. Also compiling libraries like this is rarely smooth and reading the words `FreeType comes with a sophisticated build system that is based on GNU Make` in the documentation make me cringe a bit. Thats why I thought if there is already a build system with the correct configuration on the build servers in place why not use it. It is also still worthwhile to update a 3 year old library I'd say.

**Comment by eXpl0it3r** [comment]
What did you check? Because I can't see any CMake script for any xiph library (vorbis, ogg, opus, flac). 😉

**Comment by MarioLiebisch** [comment]
Didn't know they were gone, but it basically works flawlessly as external sub projects in CMake (i.e. CMake would clone the repositories and build the libraries), except FLAC. And that's quite some deal breaker, unfortunately.

**Comment by Foaly** [comment]
Ah interesting! I thought the libraries were also built on the build servers. I think it would definitly be a good idea to invest some time into that. Because once the effort has been made to get a system like that working, it is much easier to update libraries and the barrier to do so is much lower.
Also I disagree that 3 year old libraries is fine. I think if there is an option people should always run the latest verion of any sofware. Afterall this is why updates are made 😉
Just an FYI: There is a PR waiting in the FLAC repo, that adds a CMake script (https://github.com/xiph/flac/pull/21). It's been sitting there for almost a year. Maybe we can support them, by providing a use case and maybe some testing.

Indeed compiling FreeType worked quiet easily. Using the compiled version in SFML not so much 😄 First I tried setting `SFML_USE_SYSTEM_DEPS` to true. This did not work since SFML seams to always use Frameworks on OSX. After quiet a bit of back and forth I was able to built FreeType as a framework and could get SFML to use it by disabling `SFML_USE_SYSTEM_DEPS` and simply copying the new framework into `/Library/Frameworks/`. I don't know if this is the correct way of doing this. Maybe @mantognini can shed some light on this.

Long story short: The rendering looks the same in 2.9. To me this looks like a bug or at least behaviour that differs from the documentation. If nobody is against it I would like to escalate this upstream and ask the FreeType devs if this works as intended in their opinion. Objections?

**Comment by eXpl0it3r** [comment]
> Also I disagree that 3 year old libraries is fine.

I merely pointed out that relative to SFML's release "cycle" it's not as bad as it sounds.

> I think if there is an option people should always run the latest verion of any sofware. Afterall this is why updates are made 

That's why people can always update the libraries they want. Nobody is forced to use the libraries provided by SFML, it's just that people are generally too lazy to do it themselves. 😉 

> If nobody is against it I would like to escalate this upstream and ask the FreeType devs if this works as intended in their opinion.

Please do so, if only for getting the reasoning behind the behavior. 🙂

**Comment by Foaly** [comment]
Which still doesn't make it good... 😄

Of course people can still update themselfes, but by that logic why are there any libraries bundled at all? The user could just add them themself. Like you said its about convenience. My stand is since such a convenience is offered, it should be well maintained and not risk the user shipping outdated software because he didn't check if the library he uses keeps their dependecies up to date.

Anyway onto the topic: I wrote them an email. Lets see what they say!

**Comment by mantognini** [comment]
> After quite a bit of back and forth I was able to built FreeType as a framework and could get SFML to use it by disabling `SFML_USE_SYSTEM_DEPS` and simply copying the new framework into `/Library/Frameworks/`. I don't know if this is the correct way of doing this. Maybe @mantognini can shed some light on this.

That would be a bug, probably. Alternatively, you should be able to remove the binaries & headers from SFML sources, build & install FT as a regular library (or framework) and reconfigure your cmake. At least, that's what brew is doing when installing SFML.

**Comment by Foaly** [comment]
I got in touch with a developer from FreeType and according to them there is no function they offer that does exactly what we want. So we have to implement it ourselfs. I already had a look around the code and will see if I can make a working implementation. As this might take a bit, I already send a PR for the small memory leak, so this doesn't have to wait.
